crates/node/src/txvalidation/event.rs

```diff
@@ -196,53 +198,145 @@ impl TxEvent<PropagateTx> {
+        parent_cache: &mut TxCache<Transaction<Validated>>,
+        storage: &impl ValidateStorage,
     ) -> Result<Vec<TxEvent<NewTx>>, EventProcessError> {
-        // Verify Tx'x parent is present or not.
-        let new_tx = if let Some(parent) = self.tx.payload.get_parent_tx() {
-            if cache.is_tx_cached(parent)
-                || storage
-                    .get(parent)
+        // First validate program dep
+        let prg_ok_txs = self.manage_program_dep(programid_cache, storage).await?;
+        //validate parent dep
+        let mut new_txs = vec![];
+        for prog_ok_tx in prg_ok_txs {
+            let mut ret = prog_ok_tx.manage_parent_dep(parent_cache, storage).await?;
+            new_txs.append(&mut ret);
+        }
+
+        // Transfort tx in NewTx
+        let ret = new_txs.into_iter().map(|tx| tx.into()).collect();
+
+        Ok(ret)
+    }
+
+    async fn wait_if_not_cached<'a, Kind, Fut>(
+        self,
+        cache: &mut TxCache<Kind>,
+        cache_key: std::option::Option<&'a gevulot_node::types::Hash>,
+        storage: &'a impl ValidateStorage,
+        storage_contains: impl FnOnce(&'a Hash, &'a dyn ValidateStorage) -> Fut,
+    ) -> Result<Option<Self>, EventProcessError>
+    where
+        Fut: Future<Output = eyre::Result<bool>>,
+    {
+        if let Some(cache_key) = cache_key {
+            if cache.is_tx_cached(cache_key)
+                || storage_contains(cache_key, storage)
                     .await
                     .map_err(|err| EventProcessError::StorageError(format!("{err}")))?
-                    .is_some()
             {
                 // Present return the Tx
-                Some(self)
+                Ok(Some(self))
             } else {
-                // Parent is missing add to waiting list
-                cache.add_new_waiting_tx(*parent, self);
-                None
+                // Program is missing add to waiting list
+                cache.add_new_waiting_tx(*cache_key, self);
+                Ok(None)
             }
         } else {
-            // No parent always new Tx.
-            Some(self)
-        };
+            Ok(Some(self))
+        }
+    }
+
+    // Validate if the Tx is a Run, its program has been installed.
+    // If not, wait the Tx until the program deploy Tx arrives
+    pub async fn manage_program_dep(
+        self,
+        programid_cache: &mut TxCache<Program>,
+        storage: &impl ValidateStorage,
+    ) -> Result<Vec<TxEvent<WaitTx>>, EventProcessError> {
+        // Verify that Tx associated program are present.
+        let run_tx_programs = self.tx.payload.get_run_programs_dep();
+        fn query_db_for_program<'a>(
+            cache_key: &'a Hash,
+            storage: &'a dyn ValidateStorage,
+        ) -> impl Future<Output = eyre::Result<bool>> + 'a {
+            storage.contains_program(*cache_key)
+        }
+        // Only test for the first program, because all Tx program came from the same deploy tx.
```

Comment:
This is not actually true. Nothing restricts Run transaction from combining programs from different deployments (and that is fine).

Reply:
Yes, I made a shortcut. I'll change it.

Reply:
I've updated the PR to track all program id for a Run Tx.